Goddard-Fortran-Ecosystem/pFUnit
# Enable use of add_subdirectory and build directory directly

Alternative to #251 

Fixes #250 

This PR makes several changes that allow pFUnit to be consumed in a more consistent way through different methods:

- the currently supported way, via an installation
- using the build directory directly, via `-DCMAKE_PREFIX_PATH=/path/to/pfunit/build` or `-DPFUNIT_ROOT=/path/to/pfunit/build`
- as a git submodule, via `add_subdirectory(path/to/pfunit/submodule)`
- as an external project with `FetchContent` (or `ExternalProject`)

From my testing, this shouldn't affect existing uses at all, it just allows more uses.

Here's a little snippet that can be used to switch between downloading pfunit "on demand" (download at configure time, building at build time), and an existing build or install:

```cmake
if (DOWNLOAD_PFUNIT)
  message(STATUS "Downloading pFUnit...")
  include(FetchContent)
  FetchContent_Declare(
    pfunit
    GIT_REPOSITORY https://github.com/ZedThree/pFUnit
    GIT_TAG        export-cmake-targets
    )
  # How to set options for pFUnit
  set(SKIP_OPENMP YES CACHE BOOL "" FORCE)
  FetchContent_MakeAvailable(pfunit)
else()
  find_package(PFUNIT REQUIRED)
endif()

include(CTest)
enable_testing()
set(TEST_SOURCES
  tests/always_pass.pf)
add_pfunit_test(mytests
  "${TEST_SOURCES}" "" "")
target_link_libraries(mytests mylib)
```

This PR relies on some modifications to gFTL, gFTL-shared, and fArgParse
- https://github.com/Goddard-Fortran-Ecosystem/fArgParse/pull/74
- https://github.com/Goddard-Fortran-Ecosystem/gFTL/pull/112
- https://github.com/Goddard-Fortran-Ecosystem/gFTL-shared/pull/25

so it currently points the submodules to my forks/branches -- this will need adjusting once those PRs are merged.

A brief explanation of some of the changes:
- `CMAKE_SOURCE_DIR` is the top-level project's top-level directory, so this needs to be `PROJECT_SOURCE_DIR` in order to include pfunit via `add_subdirectory`
- targets need to be exported with `export(EXPORT ...)` and installed with `install(EXPORT ...)`
- removed the `external-pfunit` interface library as it was quite painful to get the targets exported correctly. It seemed to be used in only one place, so it didn't seem like much of an advantage to keep it
- use `CMakePackageConfigHelpers` module to write the package config and version files: this removes the need for the `PFUNITConfig-version.cmake.in` file
- put `add_pfunit_test` in a standalone file, so we can use it either from an installation or via `add_subdirectory`
- add alias targets with the `PFUNIT::` namespace to make sure linking can always be done against `PFUNIT::funit` for instance
- by default, don't build the pfunit tests when built as a subproject. `CTEST_CUSTOM_PRE_TEST` ends up clobbering the main project's tests otherwise
